[Messenger] Prioritize receivers via transport configuration

petervanderwal opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | 8.2
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Issues        | 
| License       | MIT

Follow-up on #52411, which added `--all` to `messenger:consume`. A transport can now declare the order in which it is consumed:

```yaml
framework:
    messenger:
        transports:
            important: { dsn: '%env(MESSENGER_TRANSPORT_DSN)%', priority: 10 }
            async:     { dsn: '%env(MESSENGER_TRANSPORT_DSN)%' }
```

`messenger:consume --all` then consumes `important` before `async`. Higher priority comes first, the default is `0`, and the value is passed straight to the `messenger.receiver` tag, so a transport registered by a bundle can set it the same way.

Passing receiver names explicitly is unaffected: `messenger:consume async important` consumes them in the order given on the command line.

Documentation should cover: the `priority` option on `framework.messenger.transports`, that it only applies to `--all`, that higher comes first, and that it maps to the `messenger.receiver` tag for transports registered outside the framework configuration.

Commit:
[Messenger] Prioritize receivers via transport configuration

petervanderwal commented on src/Symfony/Component/Messenger/DependencyInjection/MessengerPass.php:
```diff
@@ -263,15 +265,23 @@ private function registerReceivers(ContainerBuilder $container, array $busIds):
             $receiverMapping[$id] = new Reference($id);
 
             foreach ($tags as $tag) {
+                $receiverPriority[$id] = max($tag['priority'] ?? 0, $receiverPriority[$id] ?? PHP_INT_MIN);
```
Does it happen that a receiver service has multiple `messenger.receiver` tags? If that never happens then the "take the max priority of all added tags" logic can be simplified to a single easier to read `$receiverPriority[$id] = $tag['priority'] ?? 0`.

Based on the current `FrameworkExtension` code it looked like the answer would be "no, that never happens" but I didn't know whether there were other edge-cases.

nicolas-grekas replied:
why not using PriorityTaggedServiceTrait::findAndSortTaggedServices instead?

alexislefebvre commented:
I think the changelogs should explain (briefly) how to define bigger to lower priorities, do `0` has priority over `10`?

petervanderwal commented:
> I think the changelogs should explain (briefly) how to define bigger to lower priorities, do `0` has priority over `10`?

Updated the changelog to indicate that the priority is passed on to the `messenger.receiver` service tag.

A higher priority (number) will go earlier than a lower priority -- https://symfony.com/doc/current/service_container/tags.html#tagged-services-with-priority

OskarStark commented on src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php:
```diff
@@ -1612,6 +1612,7 @@ function ($a) {
                                         ->defaultNull()
                                         ->info('Rate limiter name to use when processing messages')
                                     ->end()
+                                    ->integerNode('priority')->defaultValue(0)->info('Priority of this transport when the consumer is executed with the --all flag')->end()
```
```suggestion
                                    ->integerNode('priority')
                                        ->defaultValue(0)
                                        ->info('Priority of this transport when the consumer is executed with the --all flag')
                                    ->end()
```

bendavies commented:
this looks like a very worthwhile change. 
currently using `--all` isn't really a viable option on highly prioritized transports.

Commit:
[Messenger] Prioritize receivers via transport configuration

Transports take a "priority" option, passed on to the "messenger.receiver"
tag, which decides the order "messenger:consume --all" consumes them in.
Passing receiver names on the command line still consumes them in the order
given.
